impl/src/main/kotlin/nl/nuts/consent/bridge/listener/StateFileStorage.kt

```diff
@@ -0,0 +1,155 @@
+ */
+@Service
+class StateFileStorageControl {
+
+    @Autowired
+    lateinit var eventMetaProperties: EventMetaProperties
+
+    val stateFileStorages: MutableMap<String, StateFileStorage> = mutableMapOf()
+
+    /**
+     * Writes latest timestamp to file
+     * Precaution should be taken since this is a locking call
+     */
+    fun writeTimestamp(eventType: String, timestamp: Long) = getOrCreateStateFileStorage(eventType).writeTimestamp(timestamp)
+
+    /**
+     * Reads latest timestamp from file
+     * Precaution should be taken since this is a locking call
+     */
+    fun readTimestamp(eventType: String): Long = getOrCreateStateFileStorage(eventType).readTimestamp()
+
+    /**
+     * gets correct stateFileStorage or creates a new one if needed
+     */
+    private fun getOrCreateStateFileStorage(eventType: String) : StateFileStorage {
+        synchronized(this) {
+            var stateFileStorage = stateFileStorages[eventType]
+            if (stateFileStorage == null) {
+                stateFileStorage = StateFileStorage(eventTypeFileLocation(eventType))
+                stateFileStorages[eventType] = stateFileStorage
+            }
+            return stateFileStorage
+        }
+    }
+
+    /**
+     * returns correct File handler for given eventType, making sure it exists and is accessible
+     * must be called from synchronized context
+     */
+    private fun eventTypeFileLocation(eventType: String): File {
+        val dir = File(eventMetaProperties.location)
+        if (!dir.exists()) {
+            dir.mkdirs() // throws SecurityException is access denied
+        }
+
+        val file = dir.resolve("$eventType.$FILE_EXTENSION")
+        if (!file.exists()) {
+            file.createNewFile() // throws IOException if denied/wrong
+        }
+
+        return file
+    }
+}
+
+/**
+ * Handles writing and reading a timestamp to/from a given file with locking
+ */
+class StateFileStorage(val file: File) {
+
+    /**
+     * Writes latest timestamp to file if larger then existing timestamp
+     */
+    fun writeTimestamp(timestamp: Long) {
+        withFileLock { stream ->
+            val currentTimestamp = readTimestamp(stream)
+            if (timestamp > currentTimestamp) {
+                stream.seek(0)
+                // overwrites everything
+                stream.writeBytes("$timestamp")
+            }
+        }
+    }
+
+    /**
+     * Reads latest timestamp from file
+     */
+    fun readTimestamp(): Long {
+        var currentTimestamp = 0L
+
+        withFileLock { stream ->
+            currentTimestamp = readTimestamp(stream)
+        }
+
+        return currentTimestamp
+    }
+
+    private fun readTimestamp(stream: RandomAccessFile): Long {
+        var currentTimestamp = 0L
+
+        stream.seek(0)
+        val b = ByteArray(13)
+        try {
+            val read = stream.read(b)
+            if (read >= 13) {
+                currentTimestamp = String(b).trim().toLong()
```

Comment:
why trim it if it's 13 positions fixed length?

Reply:
Because of peoples altering the file by hand and introduce whitespace